### ISSUE TITLE
displacment Interpolation for signals with spikes applying barycenter 

### ISSUE THREAD

**Issue by sen0zk** [PythonOT/POT#223]
Hi, there
Thanks for this POT package, you have no idea how it has helped me with my job at hand. 

Currently I'm still have some trouble. I want to perform displacement interpolation between two probability distribution function with the barycenter feature. My pdf is kind of spiky, which is annoying but of great importance. I've tried the lp solver version (ot.lp.barycenter) and the entropic regularized version (ot.bregman.barycenter). The outcomes are both unsatisfactory in a way.

For the lp solver, the output (red line) suffers from fluctuation like in below:
![lp_raw_weight0 3](https://user-images.githubusercontent.com/70674898/102978569-acd79d80-453f-11eb-8e21-da1805b925f0.png)

While for the case with regularization (red dashed line), the spike is missing and the height is inappropriate according to my intuition.
![bary_bregman_weight0 3](https://user-images.githubusercontent.com/70674898/102978697-dee8ff80-453f-11eb-8eeb-3bc26d49cae4.png)

In short, I wish to figure out whether it is possible to derive the barycenter for data like this where derivative may be missing at certain point (and such feature is desirable). My background mainly resides in engineering. Thus I'm kind of confused with the underpinning mathematics.
Any help or advice is extremely welcomed.

Thanks for your attention.
Sen

**Comment by hichamjanati** [comment]
Thanks @sen0zk  we're glad you find pot useful :) 
I'm not sure how you used `lp.barycenter` it seems to work fine. If you're still having the same output with my code below please provide your code to reproduce your problem. `lp` takes a bit more time than `bregman.barycenter` but you get the spike. Note that it is normal **not** to get the spike with the regularized barycenter (bregman) which smoothens the barycenter, if you want less smoothness decrease the `reg` parameter (eps in the code below; but if too small you get numerical errors). 

( @rflamary  maybe we should add a full logsumexp version of bregman functions, I thought we had them already. The stabilized version does not fix the problem and requires changing the absorption param `tau` which is not practical. )

<img width="603" alt="Screenshot 2020-12-23 at 11 01 12" src="https://user-images.githubusercontent.com/17051500/102984681-39696800-450e-11eb-9bc5-adb3445d489c.png">


```python
import ot
import numpy as np
from matplotlib import pyplot as plt


seed = 42
rng = np.random.RandomState(seed)


def laplace(x, mean, scale):
    f = np.exp(-abs(x - mean) / scale)
    f += 1e-6  # for numerical stability
    f /= f.sum()
    return f


x = np.linspace(0., 1., 100)

a = laplace(x, 0.3, 0.05)
b = laplace(x, 0.7, 0.07)
inputs = np.stack([a, b]).T
eps = 0.001
bars = []

M = abs(x[:, None] - x[None, :]) ** 2
M /= M.max()
bar = ot.bregman.barycenter(inputs, M, eps)
bar_lp = ot.lp.barycenter(inputs, M)

plt.figure()
plt.plot(inputs, color="k")
plt.plot(bar, color="cornflowerblue", label="Bregman")
plt.plot(bar_lp, color="indianred", label="LP")
plt.legend()
plt.show()
```

**Comment by rflamary** [comment]
Sometimes LP works very well;) , @sen0zk what ground loss did you use? The LP barycenter will not work well with absolute value but work better with quadratic loss and in the code from @hichamjanati .

@hichamjanati: yes we need logsumexp implementation for sinkhorn (i will use only this one for the torch module) and barycenter, but it wont' solve the main problem that is slow (or never) convergence for small values of regularization. Still we should do a PR I agree with you.

**Comment by sen0zk** [comment]
Hi,
Thanks for both of your reply. 

I did some test on your code and found something interesting. 
1.   When the number of points in pdf vector is 100, the outcome is smooth and satisfying just like the image you posted @hichamjanati . But noises deteriorating the output signal seems to become larger as I increase the **vector length** with the lines of code:
```python
x = np.linspace(0., 1., 400)
```
And here lies the corresponding figure:
![len400](https://user-images.githubusercontent.com/70674898/103064155-49587900-45ee-11eb-9f9e-4bad71c02ba9.png)

In my case, my data got a length 1001, so my guess is that this phenomena matters more when the dimension grows. But honestly, the result shown in the above figure is acceptable. What really matters is when I try to **adjust the weight** between two ground-true pdf pattern. (since my goal is displacement interpolation) When applying the weight of [0.3, 0.7] with the length being 100, the result becomes:
![alf=0 3](https://user-images.githubusercontent.com/70674898/103064784-d6500200-45ef-11eb-9814-5ef72243f84a.png)

This then would be a problem for the application.

And I think I applied the squared  euclidean  distance for error in matrix M by running
```python
M = ot.utils.dist0(n)
```
where n is the dimension of each of my initial pdf.@rflamary

**Comment by rflamary** [comment]
Seems to me that @hichamjanati  published a potential solution to your problem:
https://arxiv.org/pdf/2006.02575.pdf

It is not yet implemented in POT but the algorithm is very simple you could try it. We should definitely add it to POT. Still i'm not sure it will keep the non differentiability. WDYT @hichamjanati ?

**Comment by hichamjanati** [comment]
I don't know why l_p bar would change like this  for non-uniform weights. Oddly, it does not take a maxiter argument, we should add one. 

@rflamary  yes you're right, It wouldn't keep the non-differentiability but it gives the right variance (i.e height) a lot faster than lp.

<img width="586" alt="Screenshot 2020-12-24 at 08 47 32" src="https://user-images.githubusercontent.com/17051500/103072157-58770100-45c5-11eb-8aa2-b06755814d9a.png">

Here its implementation and output

```python
import ot
import warnings
import numpy as np
from matplotlib import pyplot as plt


seed = 42
rng = np.random.RandomState(seed)


def laplace(x, mean, scale):
    f = np.exp(-abs(x - mean) / scale)
    f += 1e-6  # for numerical stability
    f /= f.sum()
    return f


def barycenter_debiased(P, M, eps, maxiter=5000, tol=1e-5,
                        weights=None, return_log=False):
    """Compute the Wasserstein divergence barycenter between histograms.
    """
    K = np.exp(- M / eps)
    dim, n_hists = P.shape
    bold = np.ones_like(P)
    b = bold.copy()
    c, q = np.ones((2, dim))
    Kb = K.dot(b)
    log = {'err': [], 'a': [], 'b': [], 'c': [], 'q': []}
    err = 10
    if weights is None:
        weights = np.ones(n_hists) / n_hists
    for ii in range(maxiter):
        qold = q.copy()
        a = P / Kb
        Ka = K.T.dot(a)
        q = c * np.prod((Ka) ** weights[None, :], axis=1)
        c = (c * q / K.dot(c)) ** 0.5
        Q = q[:, None]
        b = Q / Ka
        Kb = K.dot(b)
        err = abs(q - qold).max()
        if return_log:
            log["err"].append(err)
            log["a"].append(a)
            log["q"].append(q)
            log["c"].append(c)
            log["b"].append(b)
        if err < tol and ii > 10:
            break
    if ii == maxiter - 1:
        warnings.warn("*** Maxiter reached ! err = {} ***".format(err))

    if return_log:
        return q, log
    return q


x = np.linspace(0., 1., 1000)

a = laplace(x, 0.3, 0.05)
b = laplace(x, 0.7, 0.07)
inputs = np.stack([a, b]).T
eps = 0.0005
weights = np.array([0.7, 0.3])

M = abs(x[:, None] - x[None, :]) ** 2
M /= M.max()
bar_deb = barycenter_debiased(inputs, M, eps, weights=weights)

plt.figure()
plt.plot(inputs, color="k")
plt.plot(bar_deb, color="navy", label="Debiased")

plt.legend()
plt.show()
```

**Comment by rflamary** [comment]
Please do a PR @hichamjanati  when you have time, this is awesome! Also you have implemented the one from Luise for the paper it would be very interesting to have both.

About the LP it is not strictly convex so there is potential an infinite number of solutions hence the noisy solution i think (among all solutions there is not reason to keep the smooth one in LP).

**Comment by sen0zk** [comment]
I tried implementing the code @hichamjanati , the result is perfect for me. I shall go review the aforementioned paper and try to get something from it. @rflamary 
Truly appreciate your help and suggestions. Very pleasant conversation with you guys.
